documentation/04_config-options.md

Code Configuration
==================

[TOC]

Many aspects of GADGET-4 are controlled with compile-time options
rather than run-time options in the parameterfile. This is done in
order to allow the generation of highly optimised binaries by the
compiler, even when the underlying source code allows for many
different ways to run the code.  Unfortunately, this technique has the
disadvantage that different simulations typically require different
binary executables of GADGET-4, so "installing" GADGET-4 on a computer
system is not possible, and the act of compiling the code is an
integral part of working with the code. Keeping the executable at a
single place is not recommended either, because if several simulations
are run concurrently, this invokes the danger that a simulation is
started or resumed with the wrong binary. Note that while GADGET-4
checks the plausibility of some of the most important code options,
this is not done for all of them. Hence, to minimise the risk of using
the wrong code for a simulation, it is highly recommended to produce a
separate executable for each simulation that is run.

As a piece of advice, a good strategy for doing this in practice is to
create a separate directory for each simulation that is made, place a
copy of the whole simulation source code together with its makefile
into this directory, compile the code there and run it in this
directory as well, with the output directory specified as a
subdirectory of the simulation directory. In this way, the code and
its settings become a logical and integral part of the output
generated by the code. Everything belonging to a given simulation is
packaged up in one directory, and it becomes easy to reproduce what
was done even if considerable time should have passed, because the
precise version of the original code that was used and all produced
log files are readily available. An alternative is to have only a
single copy of the code, but then to use a separate directory for each
simulation that is done, including placing its configuration file
there and carrying out the compilation in this directory as well (by
passing the `make DIR=` option to the build as well). Then at least
all object files and the executable are unambiguously associated with
a particular simulation run.

The easiest way to create the file `Config.sh` needed for compilation
is to produce it as a copy of `Template-Config.sh` , and then modify
it as needed. When created from `Template-Config.sh` , the `Config.sh`
file contains a dummy list of all available compile-time code options,
with most of them commented out by default. To activate a certain
feature, the corresponding symbol should be commented in, and given
the desired value, if appropriate.

**Important Note:**

Whenever you change one of the options described below, a full
recompilation of the code may in general be necessary. For this
reason, the `Config.sh` itself has been added to the list of
dependences of each source file in the makefile, such that a complete
recompilation should happen automatically when the configuration file
is changed and the command `make` is given. Note that a manual
recompilation of the whole code can be enforced with the command `make
clean` , which will erase all object files, followed by `make` .

Most code options in `Config.sh` are switches that toggle a certain
feature on/off. Some of the symbols also take on a value if set. The
following list shows these switches with fiducial example values where
appropriate.

-------

Parallelization options                                {#parallel}
=======================



**IMPOSE_PINNING**

Ask the code to pin MPI processes to cores in an optimum fashion. This
requires the hwloc library for detecting the processor topology.  It
will generally only work on Linux. Note that most modern MPI libraries
can also be asked to arrange for the pinning via options to the MPI
start-up command, or they do this per default anyhow.

-------

**IMPOSE_PINNING_OVERRIDE_MODE**

In case the MPI start-up has already established a pinning, this is
normally detected and then `IMPOSE_PINNING` does not do
anything. Overriding this pre-established pinning can be enforced with
this option.

-------

**EXPLICIT_VECTORIZATION**

This enables a few compute kernel (currently in SPH only) to
explicitly use AVX instructions through the use of the vectorclass C++
library.

-------

**PRESERVE_SHMEM_BINARY_INVARIANCE**

This can be used to preserve the order in which partial results are
added in case the parallel tree walks use shared memory to access tree
branch data that has been imported by different processes on the same
shared memory node. In this case, exact binary invariance of output
can be retained upon reruns of the code.

-------

**SIMPLE_DOMAIN_AGGREGATION**

This tries to very roughly restrict the domain decomposition to place
adjacent domain pieces on the same shared memory node. This will then
(in some cases significantly) reduce the number of imports of
particles and nodes that need to be made, but at the price of a higher
imbalance overall. Whether this is worth it depends strongly on the
problem type. A better (forthcoming) solution will be to do the domain
decomposition hierarchically right away, taking the outline of the
shared-memory nodes into account from the outset.

-------



Basic operation mode of code                             {#basic}
============================

**PERIODIC**

Set this option if you want to have periodic boundary conditions. In
this case, the `BoxSize` parameter in the parameterfile becomes
relevant.

-------

**TWODIMS**

This effectively switches off one spatial dimension in SPH, i.e. the
code follows only 2D hydrodynamics either in the xy-, yz-, or
xz-plane. One doesn't need to tell the code explicitly which of these
planes are used, but all coordinates of the third dimension need to be
exactly equal (usually set to zero).

-------

**ONEDIMS**

Similarly to `TWODIMS`, this effectively only allows one spatial
dimension in SPH, i.e. the code follows only 1D hydrodynamics either
in the x-, y-, or z-directions. One doesn't need to tell the code
explicitly which of these directions are used, but all coordinates of
the other dimensions should be set to zero.

-------

**LONG_X_BITS** = 2

In case periodic boundary conditions are used (i.e. `PERIODIC` is on),
one can stretch the x-dimension of the box relative to `BoxSize` by
the factor `1 / 2^LONG_X_BITS`. A setting equal to 2 like in this
example, would hence mean that the boxsize in the x-direction will be
`BoxSize/4`.

-------

**LONG_Y_BITS** = 2

Similarly to the above, this switch can stretch the periodic box in the
y-direction by the factor `1/2^LONG_Y_BITS` relative to BoxSize.

-------

**LONG_Z_BITS** = 1

Finally, this option implements a possible stretch in the z-direction, similar
to `LONG_X` and `LONG_Y`. Only a subset of the stretch factors or several/all
of them may be used. The `LONG_X/Y/Z_BITS` values must be positive integers.

-------

**NTYPES** = 6

Number of particle types that should be used by the code. If not set,
a default value of 6 is adopted (which is the fixed value realised in
GADGET-2/3). The different particle types are useful as a means to
organize the simulation particles into different logical sets, which
can be helpful for analysis (e.g., one may have a type for dark
matter, one for stars, one for disk stars, etc.). The type 0 is
reserved for gas particles, and to them SPH is applied. All other
particle types are treated as collisionless particles which are
treated on the same footing as far as gravity is concerned. For each
of the types, one needs to specify a gravitational softening class in
the parameterfile.

-------

**GADGET2_HEADER**

This should be set if backwards compatibility of the snapshot file
header format with GADGET-2/3 is required. Applies only to file
formats 1 and 2. This can be useful, for example, to read in old
initial conditions. Note that in this case NTYPES may not be larger
than 6, and it should be at least as large as the number of types
actually contained in the legacy initial conditions. Also, files
cannot have more than 2^31 particles of any type in total in this
case.

-------

**SECOND_ORDER_LPT_ICS**

Processes the initial conditions before actual simulation start-up to
add in second order Lagrangian perturbation theory corrections. This
only works for specially constructed initial conditions created with
Adrian Jenkin's IC code.

-------

**LEAN**

This option is meant for special DM-only simulations that aim at high
savings in memory use. Only a uniform particle mass, a single particle
type, no hydrodynamics, and a single softening class are allowed. In
addition, one should use `TREEPM_NOTIMESPLIT`, and refrain from using
double precision to obtain a very small memory footprint.

-------

Gravity calculation                                  {#grav}
===================

**SELFGRAVITY**

This needs to be switched on if self-gravity should be computed by the
code. If it is off, one can still have hydrodynamics, and optionally a
prescribed external gravitational field.

-------

**HIERARCHICAL_GRAVITY**

If this is enabled, the time integration is carried out in a
hierarchically fashion in which the gravitational Hamiltonian is
hierarchically split into slow and fast dynamics. The advantage of
this approach is that small subsystems on short timesteps can be
cleanly decoupled from the more slowly evolving rest of the
system. This can be advantageous especially if there is a very deep and
increasingly thinly populated tail in the timestep distribution. It
also allows a time integration that is formally momentum conserving
despite the use of individual timesteps.  However, as additional
partial forces need to be computed, this approach typically entails a
somewhat higher number of force calculations. This can still be more
efficient overall, but only if the number of particles per timebin
quickly declines when going to shorter timebins.

-------

**FMM**

If this is enabled, the Fast Multipole Method for gravity calculations
instead of the one-sided classic tree algorithm is used.

-------

**MULTIPOLE_ORDER** = 2

If this is enabled, one can control the order of the multipole
expansion that is used. For a value of 3, quadrupole moments are
included in the normal tree calculation and/or in the FMM
calculations. A value of 4 includes octupole moments as well, and 5
goes further to hexadecupole moments. Note that these higher orders
increase the memory and CPU time needed for the force calculations,
but deliver more accurate forces in turn. It depends on the specific
application whether this is worthwhile. The default is a value of 2.

-------

**EXTRA_HIGH_EWALD_ACCURACY**

If this is extrapolated, the Ewald corrections are extrapolated from
the look-up table with a third order Taylor expansion (quite a bit
more expansive), otherwise a second order Taylor expansion is used.

-------

**EXTRAPOTTERM**

If this is activated, the extra multipole term in principle present
for the potential computation is evaluated, even though it does not
enter the force. For example, for monopole/dipole order (p=2), the
code will then compute quadrupole moments and use them in the
potential, but not in the force computation.

-------

**ALLOW_DIRECT_SUMMATION**

When this is set (it requires `HIERARCHICAL_GRAVITY`), the code will
calculate the gravitational forces for very small groups of particles
(the threshold for the group size is given by the constant
`DIRECT_SUMMATION_THRESHOLD`, which is set per default to 500, but one
can override this value in the config file if desired) with direct
summation. This can be useful if the timestep distribution has a tail
to very short, poorly occupied timebins. Doing the corresponding
timesteps frequently for very small sets of particles can be faster
with direct summation than doing it via the tree or FMM because then
all overhead associated with tree construction and tree walks can be
avoided.

-------

**RANDOMIZE_DOMAINCENTER**

When this is activated the whole particle set is randomly shifted in
the simulation box whenever a domain decomposition is done. This
can be useful to average out in time subtle spatial correlations in
the force errors that arise from the relative positioning of the
particle to the oct-tree geometry. Since integer coordinates are used
for particle positions, this shifting does not entail any round-off
error, and is reversible. In particular, the shifts will not be
visible in any of the outputs created. This option is basically always
recommended, and should have only positive effects.

-------

**RANDOMIZE_DOMAINCENTER_TYPES** = 2

Can be set to select one or several types (this is a bitmask) which
will then be used to locate the extension of a certain region. When
the particle set is randomly translated throughout the box, the code
will then try to avoid intersecting large oct-tree node boundaries
with this region. When this option is not set explicitely but
PLACEHIGHRESREGION is active, then this is automatically done
with a default setting RANDOMIZE_DOMAINCENTER_TYPES=PLACEHIGHRESREGION.

-------

**EVALPOTENTIAL**

When this is activated, the code also computes the gravitational
potential (which is not needed for the dynamics). This costs a bit of
extra memory and CPU-time.

-------

**TREE_NUM_BEFORE_NODESPLIT** = 4

The number of particles that may be in a tree node before it is split
into daughter nodes. If the number is reduced to 1, a fully threaded
tree is obtained in which leaf nodes contain one particle each. (Note
that empty nodes are not stored, except potentially for part of the
top-level tree if it is very finely refined.)

-------

**EXTERNALGRAVITY**

If this is switched on, an external gravitational field can be added
to the dynamics. One still has to define with different switches
and/or parameters what this external field is.

-------

**EXTERNALGRAVITY_STATICHQ**

Activates a simple external potential due to a Hernquist dark matter
halo, with parameters specified in the parameterfile.

-------


TreePM Options                                             {#treepm}
==============

**PMGRID** = 512

This enables the TreePM method, i.e. the long-range force is computed
with a PM-algorithm, and the short range force with the tree or with
FMM. The parameter has to be set to the size of the mesh that should
be used, e.g. 64, 96, 128, etc. The mesh dimensions need not
necessarily be a power of two, but the FFT is fastest for such a
choice. Note: If the simulation is not in a periodic box, then a FFT
method for vacuum boundaries is employed, where due to the required
zero-padding, only half the mesh is covering the region with
particles.  (see also the `HRPMGRID` option).

-------

**ASMTH** = 1.25

This can be used to override the value assumed for the scale that
defines the long-range/short-range force-split in the TreePM
algorithm. The default value is 1.25, in units of mesh-cells. A larger
value will make the transition region better resolved by the mesh,
yielding higher accuracy and less residual scatter in the force
matching region, but at the same time the region that needs to be
covered by the tree/FMM grows, which makes the computation more
expensive.

-------

**RCUT** = 6.0

This can be used to override the maximum radius out to which the
short-range tree-force is evaluated in case the TreePM/FMM-PM
algorithm is used. The default value is 4.5, given in
mesh-cells. Going much beyond 6 should not yield further improvements
in the way the force matching region is treated.

-------

**NTAB** = 128

This can be used to define the size of the short-range lookup table. The
default should normally be sufficient to have negligible influence on
the obtained force accuracy.

-------

**PLACEHIGHRESREGION** = 2

If this option is set (will only work together with `PMGRID` ), then
the short range force computation in the TreePM/FMM-PM algorithm is
accelerated by an additional intermediate mesh with isolated boundary
conditions that is placed onto a certain region of high-res
particles. This procedure can be useful for zoom-simulations, where
the majority of particles (the "high-res" particles) are occupying
only a small fraction of the volume. To activate this, the option
needs to be set to an integer value in the form of a bit mask that
encodes the particle type(s) that should be used to define the spatial
location of the high-resolution patch. For example, if types 1 and 4
are supposed to define this region, then the parameter should be set
to `PLACEHIGHRESREGION=2+16` , i.e.  to the sum 2^1+2^4. The actual
spatial region is then determined automatically from the current
locations of these particle types. This high-res zone may also
intersect with the box boundaries if periodic boundaries are
used. Once the spatial region is defined by the code, it however
applies to all particle types. In fact, the short range interactions
between particle pairs that fall fully inside the high-res region are
computed in terms of two contributions, one from the intermediate PM
mesh that covers the high-res region, and the other from a tree/FMM
force that however now only extends to a region that is reduced in
size (which is the origin of a possible speed-up with this
option). Particle pairs for which at least one of the partners is
outside the high-res region get the normal Tree/FMM contribution with
the standard cut-off region, corresponding to the plain TreePM
approach with the course grid covering the full simulation
volume. Also note that when the particle set is randomized throughout
the box (with the `RANDOMIZE_DOMAINCENTER` option), the code
additionally tries to avoid intersecting larger oct-tree node
boundaries by imposing certain restrictions in the randomization.


-------

**HRPMGRID** = 512

When `PMGRID` is set and non-periodic simulations are used, or when
`PLACEHIGHRESREGION` is active, the FFT algorithm is used to compute
non-periodic gravitational fields, which requires zero-padding,
i.e. only one octant of the used grid can actually be covered by the
mass distribution. The grid size that the code uses for these FFTs is
equal to `HRPMGRID` if this is set, otherwise the value of `PMGRID` is

examples/G2-neutrino/Config.sh

#########################################
#  Pick compile-time options as needed  #
#  from this template.                  #
#########################################


#--------------------------------------- Basic operation mode of code

PERIODIC                                     # enables periodic boundary condistions
#NTYPES=6                                     # number of particle types 
#RANDOMIZE_DOMAINCENTER                       # shifts the particle distribution randomly each step to reduce correlations of force errors in time
#RANDOMIZE_DOMAINCENTER_TYPES                 # can be used to set a zoom region via a particle type which is then never placed across large node boundaries
#LEAN                                         # selects a special 'lean' mode of code operation, which is optimized for aggressive memory saving
#LONG_X_BITS=2                                # can be used to reduce periodic box-dimension in x-direction relative to nominal box size
#LONG_Y_BITS=2                                # can be used to reduce periodic box-dimension in y-direction relative to nominal box size
#LONG_Z_BITS=1                                # can be used to reduce periodic box-dimension in z-direction relative to nominal box size
#GRAVITY_TALLBOX=2                            # this can be used to set-up gravity with two periodicity only in two of three dimensions
#TWODIMS                                      # restricts SPH formulation to two dimensions (for tests)
#ONEDIMS                                      # restricts SPH formulation to one dimension (for tests)
GADGET2_HEADER                               # allows reading of snapshots with GADGET-2/3 header format 
#SECOND_ORDER_LPT_ICS                         # treats second order LPT ICs generated with Adrian Jenkin's code


#--------------------------------------- Gravity calculation

SELFGRAVITY                                   # switch to enable self-gravity of particles (typically always on)
#FMM                                          # enables Fast Multipole Method instead of one-sided tree algorithm
#MULTIPOLE_ORDER=2                            # sets the multipole order of Tree or FMM computations
#EVALPOTENTIAL                                # computes gravitational potential besides force
#EXTRAPOTTERM                                 # this computes an extra multipole term for the potential which is not needed for the forces
#EXTRA_HIGH_EWALD_ACCURACY                    # this uses third-order instead of second-order Taylor expansion to interpolate Ewald corrections from table 
#ALLOW_DIRECT_SUMMATION                       # allows calculation of direct summation gravity force if only a tiny number of particles as active 
#EXTERNALGRAVITY                              # switches on inclusion of external gravitational potential
#EXTERNALGRAVITY_STATICHQ                     # example for a simple external potential due to a Hernquist halo

#--------------------------------------- TreePM Options

PMGRID=128                                    # basic mesh size for TreePM calculations
#ASMTH=1.25                                   # sets the smoothing of the PM force and thus the split-scale between Tree/FMM force and PM force
#RCUT=6.0                                     # cut-off radius beyond which Tree or FMM evaluations are stopped in TreePM / FMM-PM
#NTAB=128                                     # size of short-range look-up table
#PLACEHIGHRESREGION=2                         # bitmask selecting the particle types that are used to define the location of a secondary PM mesh
#HRPMGRID=512                                 # dimension of high-res PM grid (optional, default is HRPMGRID=PMGRID)
#FFT_COLUMN_BASED                             # uses a column-based FFT algorithm instead of the default slab-based one
#PM_ZOOM_OPTIMIZED                            # selects a communication strategy in the PM code that is better balanced for zoom simulations
#TREE_NUM_BEFORE_NODESPLIT=4                  # number of particles are are at most allowed in a tree node before it is split (can be 1)


#--------------------------------------- Time integration options

#TREEPM_NOTIMESPLIT                           # if this is activated, long-range and short-range gravity are time-integrated on a common timestep
#HIERARCHICAL_GRAVITY                         # enables hierarchical time integration of the gravity 
#FORCE_EQUAL_TIMESTEPS                        # this chooses a global timestep for all particles


#--------------------------------------- Treatment of gravitational softening

#INDIVIDUAL_GRAVITY_SOFTENING=4+8+16+32       # bitmasks which selects the particle type(s) which pick their softening class based on particle mass
#NSOFTCLASSES=4                               # number of different softening classes
#ADAPTIVE_HYDRO_SOFTENING                     # makes SPH gas particles pick an adaptive gravitational softening proportional to their SPH smoothing lengths


#--------------------------------------- SPH treatmeant and formulation

#REUSE_HYDRO_ACCELERATIONS_FROM_PREVIOUS_STEP # does not recompute the pressure forces after application of source functions 
#VISCOSITY_LIMITER_FOR_LARGE_TIMESTEPS        # limits the acceleration due to the viscosity  
#PRESSURE_ENTROPY_SPH                         # enables the Hopkins (2013) pressure-entropy formulation, other density-entropy is used
#GAMMA=1.4                                    # sets the adiabatic index
#ISOTHERM_EQS                                 # selects an isothermal equation of state
#IMPROVED_VELOCITY_GRADIENTS                  # use higher-order gradients of the velocities according to Hu et. al (2014)


#--------------------------------------- SPH kernels

#CUBIC_SPLINE_KERNEL                          # uses the cubic spline kernel (default)
#WENDLAND_C2_KERNEL                           # use the Wendland C2 kernel from Dehnen & Aly 2012
#WENDLAND_C4_KERNEL                           # use the Wendland C4 kernel from Dehnen & Aly 2012
#WENDLAND_C6_KERNEL                           # use the Wendland C6 kernel from Dehnen & Aly 2012
#WENDLAND_BIAS_CORRECTION                     # reduces self-contribution for Wendland kernels according to Dehnen & Aly 2012


#--------------------------------------- SPH viscosity options

#TIMEDEP_ART_VISC                             # enables time dependend viscosity
#HIGH_ART_VISC_START                          # start with high rather than low viscosity
#NO_SHEAR_VISCOSITY_LIMITER                   # turns of the shear viscosity supression


#--------------------------------------- Extra physics

#COOLING                                      # Enables radiative atomic cooling by hydrogen and helium
#STARFORMATION                                # Enables star formation with the Springel & Hernquist (2003) model


#---------------------------------------- Single/double precision and data types

#POSITIONS_IN_32BIT                           # if set, use 32-integers for positions  (default for single precision)
#POSITIONS_IN_64BIT                           # if set, use 64-integers for positions  (default for double precision)
#POSITIONS_IN_128BIT                          # if set, use 128-integers for positions (may only work on some systems)
#DOUBLEPRECISION=1                             # if activated and set to 1, use double precision internally, for 2 use mixed precision, otherwise single precision
#DOUBLEPRECISION_FFTW                         # if set, carries out FFTs in double precision, otherwise single precision
#USE_SINGLEPRECISION_INTERNALLY               # reduces default double precision for most internal computations to single precision
#ENLARGE_DYNAMIC_RANGE_IN_TIME                # eextend the dynamic range of the integer timeline from 32 to 64 bits
#IDS_32BIT                                    # selects 32-bit IDs for internal storage (default)
#IDS_48BIT                                    # selects 48-bit IDs for internal storage
#IDS_64BIT                                    # selects 64-bit IDs for internal storage
#OUTPUT_IN_DOUBLEPRECISION                    # output files will be written in double precision


#---------------------------------------- Output/Input options

#INITIAL_CONDITIONS_CONTAIN_ENTROPY
#OUTPUT_VELOCITY_GRADIENT                     # output velocity gradients
#OUTPUT_PRESSURE                              # output gas pressure   
#OUTPUT_ENTROPY                               # output gas entropy
#OUTPUT_CHANGEOFENTROPY                       # output rate of change of entropy
#OUTPUT_POTENTIAL                             # output gravitational potential
#OUTPUT_ACCELERATION                          # output total acceleration
#OUTPUT_TIMESTEP                              # output timesteps
#OUTPUT_DIVVEL                                # output velocity divergence
#OUTPUT_CURLVEL                               # output velocity curl
#OUTPUT_COOLHEAT                              # output actual energy loss/gain in cooling/heating routine
#OUTPUT_PRESSURE_SPH_DENSITY                  # output also density computed in the pressure-entropy forumulation
#OUTPUT_VISCOSITY_PARAMETER                   # output current viscosity parameter in SPH particle output (only for time-dependent viscosity)
#OUTPUT_NON_SYNCHRONIZED_ALLOWED              # allow snapshot creation also at steps that are not fully synchronized
#OUTPUT_VELOCITIES_IN_HALF_PRECISION          # special option to store velocities in reduced precision
#OUTPUT_ACCELERATIONS_IN_HALF_PRECISION       # special option to store acclerations in reduced precision
#OUTPUT_COORDINATES_AS_INTEGERS               # special option to store coordinates as integers that are used internally
#POWERSPEC_ON_OUTPUT                          # computes a matter power spectrum when the code writes a snapshot output
#ALLOW_HDF5_COMPRESSION                       # applies HDF5 loss-less compression to selected output fields
#REDUCE_FLUSH                                 # do not flush the I/O streams of the log-files every system step


#---------------------------------------- On the fly FOF groupfinder

#FOF                                          # enable FoF output
#FOF_PRIMARY_LINK_TYPES=2                     # bitmask, 2^type for the primary dark matter type
#FOF_SECONDARY_LINK_TYPES=1+16+32             # bitmask, 2^type for the type(s) linked to nearest primary particle
#FOF_GROUP_MIN_LEN=32                         # minimum group length (default is 32)
#FOF_LINKLENGTH=0.2                           # linking length for FoF (default=0.2)
#FOF_ALLOW_HUGE_GROUPLENGTH                   # set if there are groups and subhalos with more than 2 billion particles in length


#---------------------------------------- Subfind

#SUBFIND                                      # enables substructure finder
#SUBFIND_HBT                                  # use previous subhalo catalogue instead of density excursions to define subhalo candidates
#SUBFIND_STORE_LOCAL_DENSITY                  # calculates local densities and velocity dispersions for all particles and stores them in snapshots
#SUBFIND_ORPHAN_TREATMENT                     # creates special snapshots with formerly most bound particles


#---------------------------------------- Merger tree code

#MERGERTREE                                   # enables on-the-fly calculation of descendants, and merger tree construction from group catalogues


#---------------------------------------- On-the-fly lightcone creation

#LIGHTCONE                                    # master option for lightcone output creation
#LIGHTCONE_PARTICLES                          # produces particle lightcones
#LIGHTCONE_MASSMAPS                           # produces mass shells on the lightcone
#LIGHTCONE_PARTICLES_GROUPS                   # computes groups for particles buffered on the lightcone
#LIGHTCONE_OUTPUT_ACCELERATIONS               # stores accelerations for particles on lightcone
#LIGHTCONE_IMAGE_COMP_HSML_VELDISP            # option for computing densities and smoothing length for lightcones in postprocessing
#REARRANGE_OPTION                             # special option to reorder lightcone data in mergertree order


#--------------------------------------- IC creation
 
#NGENIC=256                                   # generate cosmological ICs, set NGENIC to the FFT grid size used for IC generation
#NGENIC_2LPT                                  # applies 2LPT instead of just Zeldovich approximation
#CREATE_GRID                                  # start with a regular Cartesian DM particle grid, instead of reading a glass file (for NGENIC)
#GENERATE_GAS_IN_ICS                          # add SPH particles to created or read dark matter only ICs
#SPLIT_PARTICLE_TYPE=4+8                      # specifies particle types to be split if GENERATE_GAS_IN_ICS is activated
#NGENIC_FIX_MODE_AMPLITUDES                   # when activated, this leaves the mode amplitudes at sqrt(P(k)), instead of sampling from a Rayleigh distribution
#NGENIC_MIRROR_PHASES                         # if this is activated, all phases are turned by 180 degrees
#NGENIC_TEST                                  # can be used to create ICs, measure the power spectrum, and then stop


#----------------------------------------Parallelization options

#IMPOSE_PINNING                               # enables pinning of MPI processes to CPU cores
#IMPOSE_PINNING_OVERRIDE_MODE                 # tries to do the pinning even if a prior pinning is detected
#PRESERVE_SHMEM_BINARY_INVARIANCE             # preserve binary invariance of results despite machine weather, at the price of more tree walk overhead 
#EXPLICIT_VECTORIZATION                       # use AVX at selected places in SPH kernels through the vectorclass C++ library
#SIMPLE_DOMAIN_AGGREGATION                    # this is an experimental modification of the domain decomposition algorithm (can either help or harm performance)


#---------------------------------------- MPI related settings

#NUMBER_OF_MPI_LISTENERS_PER_NODE=1           # set such that the number of MPI-ranks per node and listener is maller than MAX_NUMBER_OF_RANKS_WITH_SHARED_MEMORY 
#MAX_NUMBER_OF_RANKS_WITH_SHARED_MEMORY=64    # default is 64, but can also be set to 32
#NUMPART_PER_TASK_LARGE                       # set this if the number of particles per task is so large that more than 2 GB are comprised just by particle data
#USE_MPIALLTOALLV_IN_DOMAINDECOMP             # replaces hypercube communication in domain particle exchance with a single MPI_Allgatherv (can be less stable)
#MPI_HYPERCUBE_ALLGATHERV                     # if your MPI-library uses too much internal storage for MPI_Allgatherv, this uses a hypercube as a work-around
#MPI_MESSAGE_SIZELIMIT_IN_MB=200              # limit the message size of very large MPI transfers
#ISEND_IRECV_IN_DOMAIN                        # uses asynchronous communication instead of synchronous communication in hypercube pattern (can be less stable)


#---------------------------------------- Testing and Debugging options

#HOST_MEMORY_REPORTING                        # reports after start-up the available system memory by analyzing /proc/meminfo (active by default on Linux)
#ENABLE_HEALTHTEST                            # can be used to enable some auto-checking of the machine you're running on
#DEBUG                                        # enables core-dumps in case MPI_Init() has disabled them
#DEBUG_ENABLE_FPU_EXCEPTIONS                  # tries to enable FPU exceptions so that a bad floating point operation triggers a core dump
#DEBUG_SYMTENSORS                             # carries out a few unit tests related to the tensor algebra routines of the code
#FORCETEST=0.001                              # calculates for given fraction of particles direct summation forces to check accuracy of tree force
#FORCETEST_FIXEDPARTICLESET                   # if this set, always the same particle receive a force accuracy check during a run
#FORCETEST_TESTFORCELAW=1                     # special option for measuring the effective force law, can be set to 1 or 2 for TreePM+HighRes
#VTUNE_INSTRUMENT                             # outputs auxiliary info to simplify Vtune performance analysis
#DEBUG_MD5                                    # make MD5 check sum routines available for debugging and code development
#SQUASH_TEST                                  # special check-option for code development
#DOMAIN_SPECIAL_CHECK                         # special check-option for code development
#RECREATE_UNIQUE_IDS                          # before carrying out the ID uniqueness test, this sets new IDs (e.g. to fix broken ICs)
#NO_STOP_BELOW_MINTIMESTEP                    # do not stop when the code wants to go below a prescribed minimum timestep
#TILING=2                                     # duplicate an IC in each dimension for scaling tests
#DO_NOT_PRODUCE_BIG_OUTPUT                    # for scaling tests, disable the writing of restart files and snapshot dumps
#MEASURE_TOTAL_MOMENTUM                       # special test option for code development
#EWALD_TEST                                   # a development test for the Ewald tables
#STOP_AFTER_STEP=10                           # ends a simulation after the specified timestep (to simplify scalability tests)
#TREE_NO_SAFETY_BOX                           # when set, this disables the geometric 'near node' protection


#---------------------------------------- Neutrino 

NEUTRINO
#STERILE
